### PR TITLE
SR4R Redesign: Make `RejectionReasonsHistory` work new R4R data

### PR DESCRIPTION
**Pull request by steventux**
## Context

We show candidates who are applying again the feedback they have received from their previous rejected application.
This feedback is specifically the personal statement and subject knowledge reasons for rejection.
The redesigned SR4R form and JSON format stores this information in a different way to we need to be able to work with both formats when presenting the candidate with this information.

## Changes proposed in this pull request

- Adds ability for `RejectionReasonsHistory` to present new R4R data.
- Adds conditional logic around 'Other' details.

## Link to Trello card

https://trello.com/c/lDefSjFC/4889-sr4r-redesign-candidate-apply-2-review-presentation-when-previous-application-was-rejected

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

**Comment by cwrw** [comment]
Received @adamsilver's seal of approval

File: spec/models/candidate_interface/rejection_reasons_history_spec.rb [hunk 1]
```diff
@@ -35,12 +35,105 @@
...
+                    },
+                  },
+                ],
+              },
+            ],
+          }
+        end
+
+        it 'returns the related rejections in history items' do
+          feedback = described_class.all_previous_applications(current_application_form, :becoming_a_teacher)
```
Comment (SamJamCul) [inline spec/models/candidate_interface/rejection_reasons_history_spec.rb:91]: Rename to `history_items`

File: app/components/candidate_interface/review_rejection_reasons_component.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    include ViewHelper
+
+    attr_reader :rejection_reasons
+
+    def initialize(rejection_reasons)
+      @rejection_reasons = rejection_reasons
+    end
+
+    def hide_other_label?
+      rejection_reasons.count == 1 && rejection_reasons.first.label == 'Other'
```
Comment (carlosmartinez) [inline app/components/candidate_interface/review_rejection_reasons_component.rb:12]: some pleasingly obscure rails magic: `rejection_reasons.one?`

File: app/models/candidate_interface/rejection_reasons_history.rb [hunk 3]
```diff
@@ -49,19 +49,46 @@ def collect_previous_applications(application_form)
 
     def extract_history(application_choices)
       application_choices.includes(:provider).filter_map do |choice|
-        reasons_for_rejection = ReasonsForRejection.new choice.structured_rejection_reasons
-        feedback = reasons_for_rejection.send reasons_for_rejection_method
+        next if choice.structured_rejection_reasons.blank?
+
+        feedback = feedback_for_choice(choice)
 
         if feedback.present?
```
Comment (carlosmartinez) [inline app/models/candidate_interface/rejection_reasons_history.rb:56]: could consolidate the filter conditions: 
```
if feedback.present? && choice.structured_rejection_reasons.present?
```

File: spec/components/candidate_interface/review_rejection_reasons_component_spec.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  describe '#hide_other_label?' do
+    it 'returns true if there is a single other rejection reasons' do
+      personal_statement_rejection_details = { id: 'personal_statement_other_details', text: 'Personal statement Bad' }
+      rejection_reasons = [
+        RejectionReasons::Reason.new(id: 'personal_statement_other', label: 'Other', details: personal_statement_rejection_details),
+      ]
+
+      result = described_class.new(rejection_reasons)
+
+      expect(result.hide_other_label?).to be(true)
```
Comment (carlosmartinez) [inline spec/components/candidate_interface/review_rejection_reasons_component_spec.rb:48]: `expect(result).to hide_other_label`

File: spec/components/candidate_interface/review_rejection_reasons_component_spec.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    it 'returns false if there is a single non other details rejection reasons' do
+      rejection_details = { id: 'quality_of_writing_details', text: 'Quality Bad' }
+      rejection_reasons = [RejectionReasons::Reason.new(id: 'quality_of_writing', label: 'Quality of writing', details: rejection_details)]
+
+      result = described_class.new(rejection_reasons)
+
+      expect(result.hide_other_label?).to be(false)
```
Comment (carlosmartinez) [inline spec/components/candidate_interface/review_rejection_reasons_component_spec.rb:57]: `expect(result).not_to hide_other_label`

File: spec/components/candidate_interface/review_rejection_reasons_component_spec.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+      qualitfy_rejection_details = { id: 'quality_of_writing_details', text: 'Quality Bad' }
+      personal_statement_rejection_details = { id: 'personal_statement_other_details', text: 'Personal statement Bad' }
+      rejection_reasons = [
+        RejectionReasons::Reason.new(id: 'quality_of_writing', label: 'Quality of writing', details: qualitfy_rejection_details),
+        RejectionReasons::Reason.new(id: 'personal_statement_other', label: 'Other', details: personal_statement_rejection_details),
+      ]
+
+      result = described_class.new(rejection_reasons)
+
+      expect(result.hide_other_label?).to be(false)
```
Comment (carlosmartinez) [inline spec/components/candidate_interface/review_rejection_reasons_component_spec.rb:70]: `expect(result).not_to hide_other_label`

File: spec/models/candidate_interface/rejection_reasons_history_spec.rb [hunk 7]
```diff
@@ -35,12 +35,105 @@
...
+
+          history_item = feedback.first
+          expect(feedback.count).to eq(1)
```
Comment (carlosmartinez) [inline spec/models/candidate_interface/rejection_reasons_history_spec.rb:94]: readability option: `expect(feedback).to have(1).item`

**Review by carlosmartinez — APPROVED** [review]

File: spec/components/candidate_interface/review_rejection_reasons_component_spec.rb (hunk 4 above)
Comment (cwrw) [reply spec/components/candidate_interface/review_rejection_reasons_component_spec.rb:48]: Attempting to use predicate matchers yields a `NameError: undefined local variable or method `hide_other_label'` error due to it not being prepended by 'be_` or `have_` (see the [matcher regex](https://github.com/rspec/rspec-expectations/blob/v3.11.0/lib/rspec/matchers.rb#L960))which doesn't read well with this method. I think using `be(true)` here is not risky as this method returns a boolean. Let me know your thoughts

File: spec/models/candidate_interface/rejection_reasons_history_spec.rb (hunk 7 above)
Comment (cwrw) [reply spec/models/candidate_interface/rejection_reasons_history_spec.rb:94]: Unfortunately the `have` matchers are part of the `rspec-collection_matchers` gem (which is not included in this repo) and have been removed from RSpec 3

**Comment by cwrw** [comment]
Thanks @carlosmartinez and @SamJamCul for the review. I've addressed and responded to all comments if you could have another look

File: spec/components/candidate_interface/review_rejection_reasons_component_spec.rb (hunk 4 above)
Comment (carlosmartinez) [reply spec/components/candidate_interface/review_rejection_reasons_component_spec.rb:48]: oh right, sorry, i should've checked it locally before requesting. happy with your suggestion.

**Comment by carlosmartinez** [comment]
> Thanks @carlosmartinez and @SamJamCul for the review. I've addressed and responded to all comments if you could have another look

**Pull request merged**